scripts/generate_release_notes.groovy

```diff
@@ -0,0 +1,272 @@
+    newRulesXml = new XmlSlurper().parse(newRulesFile)
+} catch (Exception e) {
+    writer.close()
+    println "Error parsing XML files: ${e.message}"
+    System.exit(1)
+}
+
+// Extract rule keys from old rules (they are attributes)
+def oldRules = []
+oldRulesXml.rule.each { rule ->
+    // Skip commented out rules
+    if (rule.@key.toString()) {
+        oldRules << [
+            key: rule.@key.toString(),
+            configKey: rule.configKey.text(),
+            priority: rule.priority.text(),
+            status: rule.status.text()
+        ]
+    }
+}
+
+// Extract rule keys from new rules (they are elements)
+def newRules = []
+newRulesXml.rule.each { rule ->
+    newRules << [
+        key: rule.key.text(),
+        internalKey: rule.internalKey.text(),
+        severity: rule.severity.text(),
+        name: rule.name.text(),
+        status: rule.status.text()
+    ]
+}
+
+// Function to extract alternative rule references from markdown files
+def extractAlternativeRule(ruleKey) {
+    def mdFile = new File("docs/rules/${ruleKey}.md")
+    if (!mdFile.exists()) {
+        return ""
+    }
+
+    def content = mdFile.text
+    def pattern = Pattern.compile(":warning: This rule is \\*\\*deprecated\\*\\* in favour of ((?:\\[[^\\]]+\\]\\([^)]+\\)|`[^`]+`)(?:, (?:\\[[^\\]]+\\]\\([^)]+\\)|`[^`]+`))*)")
+    def matcher = pattern.matcher(content)
+
+    if (matcher.find()) {
+        def alternatives = matcher.group(1)
+        // Extract all alternatives from the matched group, preserving the original markdown format
+        def altPattern = Pattern.compile("(\\[[^\\]]+\\]\\([^)]+\\))|((`[^`]+`))")
+        def altMatcher = altPattern.matcher(alternatives)
+        def result = []
+        while (altMatcher.find()) {
+            result << (altMatcher.group(1) ?: altMatcher.group(3))
+        }
+        return result.join(", ")
+    }
+
+    return ""
+}
+
+// Get sets of rule keys for comparison
+def oldRuleKeys = oldRules.collect { it.key }
+def newRuleKeys = newRules.collect { it.key }
+
+// Find rules that are in old but not in new
+def removedRules = oldRules.findAll { !newRuleKeys.contains(it.key) }
+
+// Find rules that are in new but not in old
+def addedRules = newRules.findAll { !oldRuleKeys.contains(it.key) }
+
+// Find rules that exist in both
+def commonRuleKeys = oldRuleKeys.intersect(newRuleKeys)
+def commonRules = commonRuleKeys.collect { key ->
+    def oldRule = oldRules.find { it.key == key }
+    def newRule = newRules.find { it.key == key }
+    def alternative = extractAlternativeRule(key)
+    [
+        key: key,
+        oldConfigKey: oldRule.configKey,
+        newInternalKey: newRule.internalKey,
+        oldPriority: oldRule.priority,
+        newSeverity: newRule.severity,
+        name: newRule.name,
+        oldStatus: oldRule.status,
+        newStatus: newRule.status,
+        alternative: alternative
+    ]
+}
+
+// Generate report
+writer.writeLine("# PMD Rules Release Notes for version 4.1.0")
```

Comment:
can this version 4.1.0 be dynamic instead of hardcoded and easy to forget to change?